src/http/compose-middleware.js

```diff
@@ -0,0 +1,51 @@
+import { ensureAsync, asyncReduce } from '../utils'
+import { merge } from 'lodash/fp'
+
+/**
+ *
+ * A utility function that composes multiple request middlewares into a single function.
+ * This can be used to create more complex `before` hooks for {@link http}.
+ *
+ * @name composeMiddleware
+ * @type Function
+ * @param {...Function} middlewares - Functions that receive and return request options
+ * @returns {Function} A composed middleware function
+ *
+ * @example
+ *
+ * function addBearerToken () {
+ *    const token = getTokenFromStorage()
+ *    if (token) return { bearerToken: token }
+ * }
+ * 
+ * function addPathToEndpoint ({ endpoint }) {
+ *    return {
+ *      endpoint: endpoint + '/some-path'
+ *    }
+ * }
+ * 
+ * const before = composeMiddleware(
+ *    addBearerToken,
+ *    addPathToEndpoint,
+ * )
+ * 
+ * // this will call both middlewares before the request
+ * http('/users', { before }) 
+ * 
+ * 
+ */
+
+function composeMiddleware (...middlewares) {
+  // Note: An async reducer is used because we want to allow middleware
+  // to return promises if they want.
+  const asyncMiddlewares = middlewares.map(ensureAsync)
+  return function callMiddleware (initialValue) {
+    return asyncReduce(
+      asyncMiddlewares,
+      (acc, middleware) => middleware(acc).then(result => merge(acc, result || {})),
```

Comment:
Is it worth trying to safeguard against keys with an undefined value here? That probably breaks Single Responsibility, but would help if users forget to add guard clauses to each individual middleware

Reply:
I don't think so. The way we've used undefined keys before has been a work-around to handle the use case that guard clauses provide- if someone returns a key with an undefined value from middleware, I'd assume they want it to be passed in.

Reply:
LMK if that idea makes sense or if you'd like to appeal 😄

Reply:
👍 Yep! That's where I was leaning but wanted to check. Let's not bastardize it just for my use case 😏